### PR TITLE
feat: add agent injector telemetry

### PR DESCRIPTION
**Pull request by gsantos-hc**
Add Prometheus metrics to monitor the Agent Injector's performance. New metrics include a gauge of current requests being processed by the webhook, a summary of request processing times, and a count of successful and failed injections by Kubernetes namespace.

Fixes AG-005161.

### CODE REVIEW

File: subcommand/injector/command.go
Comment (gsantos-hc) [inline subcommand/injector/command.go:235]: Note that this ultimately calls `registry.MustRegister`, which will [panic](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#MustRegister) on failure. The behavior matches [what's done in VSO](https://github.com/hashicorp/vault-secrets-operator/blob/1273434f78c5a2561ec461715306d146cc13fe69/main.go#L366) but wanted to call it out just in case.

File: agent-inject/metrics.go
Comment (gsantos-hc) [inline agent-inject/metrics.go:26]: Went with a fixed set of buckets, similar to [what's done for VSO](https://github.com/hashicorp/vault-secrets-operator/blob/1273434f78c5a2561ec461715306d146cc13fe69/vault/client_metrics.go#L24), but we could also use something like [ExponentialBuckets](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#ExponentialBuckets)?

**Review by tvoran — COMMENTED** [review]
This worked well when I was running it locally, just left a couple thoughts.

File: agent-inject/metrics.go
Comment (gsantos-hc) [reply agent-inject/metrics.go]: Fixed in 64edf1d6c7ee8173b35f109cc4f5321542e309a4.

File: agent-inject/handler.go
Comment (gsantos-hc) [reply agent-inject/handler.go]: Fixed in 632154f496aaf545920f959739469ac735e09ae5

File: agent-inject/metrics.go
Comment (gsantos-hc) [reply agent-inject/metrics.go]: Decided to stick with a predefined set of buckets but updated the configuration to reduce their number (reducing cardinality) and increase the highest-tier bucket to 10 seconds (from 1 second).

File: agent-inject/handler.go
Comment (tvoran) [inline agent-inject/handler.go:154]: We may also want to call `incrementInjectionFailures()` for the error returns further up in this function too? But this should suffice as-is; I could even see those being a different error metric added in the future.

**Review by tvoran — DISMISSED** [review]
Thanks!

**Comment by gsantos-hc** [comment]
Force-pushed:
* Rebase from the latest `main`
* Add a breakdown of injection types (normal sidecar vs initContainer-only) to further help identify potential targets for migration to VSO

File: agent-inject/handler.go
Comment (tvoran) [inline agent-inject/handler.go:29]: Feels like this could get noisy for folks who aren't looking to move to VSO just yet? And workloads other than init-only are certainly suitable for VSO too. I think I'd prefer to leave this sort of warning out, please.

File: agent-inject/handler.go
Comment (tvoran) [inline agent-inject/handler.go:273]: Instead of adding onto the Warning that's sent back to the user here, I'd suggest changing Mutate() to return a struct that contains the AdmissionResponse along with metadata fields for what was injected, i.e. a couple of bool's for `init` and `sidecar` injections. Something like:

```go
type MutateResponse struct {
	InjectedInit    bool
	InjectedSidecar bool
	Resp            *admissionv1.AdmissionResponse
}
```

That would also make it easier for the metrics to show the other injection case of sidecar only.

Alternately we could perhaps set the injected container info in the AuditAnnotations map on the AdmissionResponse.

File: agent-inject/metrics.go
Comment (tvoran) [inline agent-inject/metrics.go:16]: The injector adds an init container and a sidecar container by default, so default = "sidecar" doesn't sound right to me. I'd say there are three injection types: init only, sidecar only, and both. In the future we'll hopefully support [native sidecars](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) as an injection type.

File: agent-inject/metrics.go
Comment (tvoran) [inline agent-inject/metrics.go:54]: Seems like we might as well account for the sidecar-only injection type too?

File: agent-inject/handler.go
Comment (gsantos-hc) [reply agent-inject/handler.go:273]: Sounds good. I looked into AuditAnnotations, but since those would get persisted to Kubernetes audit logs decided to go with Warnings instead.

Agreed that it's better to just change the signature of `Mutate()` and return the metadata we need for telemetry.

File: agent-inject/metrics.go
Comment (tvoran) [inline agent-inject/metrics.go:12]: I doubt we'll change the default anytime soon, but it might be more clear to name this label "both" or "init_and_sidecar".

**Review by tvoran — COMMENTED** [review]
This is looking great, just a couple minor thoughts (but nothing to hold this up over).

I was also playing around with ways to test the logic in incrementInjections(), and it looks like we could use prometheus' testutil something like this: https://gist.github.com/tvoran/b6ec51e716fd10efe7bb1847def8e98a

File: agent-inject/handler.go
Comment (benashz) [inline agent-inject/handler.go:93]: I wonder if would be worth tracking the failures as well. We do something like this in VSO here: https://github.com/hashicorp/vault-secrets-operator/blob/6ab056c22acb5dc4812620233e3b141fb9e02f9c/vault/client.go#L810

**Review by benashz — COMMENTED** [review]
Looking good. I have included some additional feedback for your consideration. Thanks!

File: agent-inject/metrics.go
Comment (gsantos-hc) [reply agent-inject/metrics.go:4]: Fixed in a095cc4e68419b68f78fffdd0a74617de6c929ad

File: agent-inject/handler.go
Comment (gsantos-hc) [reply agent-inject/handler.go:93]: Makes sense, and it would unblock @tvoran's feedback [here](https://github.com/hashicorp/vault-k8s/pull/703#discussion_r1847772903).

@benashz Given that it goes beyond the scope we were asked for, is it OK if I push that to a follow-on PR?

File: agent-inject/metrics.go
Comment (benashz) [inline agent-inject/metrics.go]: I think we should include the help message, just drop the extra "Histogram" bit.

**Pull request merged**
